raysect/core/math/function/vectorfunction2d/autowrap.pyx

```diff
@@ -0,0 +1,92 @@
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+import numbers
+from raysect.core.math.vector cimport Vector3D
+from raysect.core.math.function.vectorfunction2d.base cimport VectorFunction2D
+from raysect.core.math.function.vectorfunction2d.constant cimport ConstantVector2D
+
+
+cdef class PythonVectorFunction2D(VectorFunction2D):
+    """
+    Wraps a python callable object with a VectorFunction2D object.
+
+    This class allows a python object to interact with cython code that requires
+    a VectorFunction2D object. The python object must implement __call__() expecting
+    two arguments.
+
+    This class is intended to be used to transparently wrap python objects that
+    are passed via constructors or methods into cython optimised code. It is not
+    intended that the users should need to directly interact with these wrapping
+    objects. Constructors and methods expecting a VectorFunction2D object should be
+    designed to accept a generic python object and then test that object to
+    determine if it is an instance of VectorFunction2D. If the object is not a
+    VectorFunction2D object it should be wrapped using this class for internal use.
+
+    See also: autowrap_vectorfunction2d()
+
+    :param object function: the python function to wrap, __call__() function must
+    be implemented on the object.
+    """
+    def __init__(self, object function):
+        self.function = function
+
+    cdef Vector3D evaluate(self, double x, double y):
+        return self.function(x, y)
+
+
+cdef VectorFunction2D autowrap_vectorfunction2d(object function):
+    """
+    Automatically wraps the supplied python object in a PythonVectorFunction2D or ConstantVector2D object.
+
+    If this function is passed a valid VectorFunction2D object, then the
+    VectorFunction2D object is simply returned without wrapping.
+
+    If this function is passed a Vector3D, a ConstantVector2D object is
+    returned.
+
+    This convenience function is provided to simplify the handling of
+    VectorFunction2D and python callable objects in constructors, functions and
+    setters.  """
+
+    if isinstance(function, VectorFunction2D):
+        return <VectorFunction2D> function
+    try:
+        function = Vector3D(*function)
```

Comment:
To be fair, `function` is a pretty bad name for a variable here. At this point, we're trying to autowrap a constant which can be converted to a `Vector3D`, i.e. either an actual `Vector3D` instance or an iterable of length 3 such as a tuple. This is based on similar code in `function2d.autowrap`, which calls `Constant2D(function)`.

I could either add a comment explaining what is going on here, or introduce an alias to make things clearer, e.g.
```python
iterable = function
try:
    vector = Vector3D(*iterable)
except (TypeError, ValueError):
    return PythonVectorFunction2D(function)
else:
    return ConstantVector2D(vector)
```
Do you have a preference?

Reply:
Oh I see. I think it best to replace `function` with `obj` here and treat it as an arbitrary object.

It occurs to me that this would be another place we should be detecting raysect functions and callables:

```
If it is a compatible raysect function -> simply return it.
If it is an incompatible raysect function -> error
If it is a callable -> python function
If is is a Vector3D -> constant
Everything else try as a sequence -> if that fails throw an exception reporting that autowrapping/coercion failed.
```

See my other comment on base classes for the function framework for how this can be accomplished.